Support for expired passwords and pam message localization (#472)

hoehnth opened the pull request:
hi folks,

because we required the feature here urgently, i have implemented password change for expired passwords in sddm (solves issue #472) and also pam message localization (i18n) using system locale.
i tried to follow the sddm concept, and propagate the pam conversation and user responses through backend/daemon/greeter forth and back.
two new components for password change (PasswordChange.qml, PamConvHelper.qml), some new signals and Main.qml properties were added.
password change dialog may need some polishing. have prepared the sddm stock themes for pwd change. to include the feature in new themes is simple.

implementation notes:

first i considered to handle password change directly in greeter (via Auth.cpp), because Auth is prepared for qml interaction.
but this would interrupt the pam conversation in the backend (which handles user session with e.g. special group permissions).
so i decided to route pam conversation through all the separate procs: backend, daemon, greeter (and intermediate sockets) like for loginFailed message.
drawback: new password is exposed in ram on its way propagated through greeter, daemon, helper, (pam) backend and sockets. so have to clear pwd after use.
then instead of sending back a complete request with user responses to backend i decided to just propagate the new password back to keep it simple.
the password changing conversation with greeter is prepared for requests with AuthPrompt.CHANGE_PASSWORD prompts.
i tested only for pam backend. password backend should also work but not used so far.
i separated Request and Prompt and its stream operators out of AuthMessages into AuthBase.
this is convenient for request/response transport between (pam) backend and GreeterProxy (via SocketServer).

regards,
thomas

done:

- PasswordChange dialog with Loader, loading on demand
- localization (i18n) for pam messsages and prompts
- system locale setting (locale.conf) for pam backend
- pam conversation transport between pam backend and greeter
- pam results enum for evaluation in themes (e.g. PAM_MAXTRIES)
- sddm stock themes prepared for password change (builtin maui, elarun, maldives) 
- readme how to use password change in new themes

todo:

- test password change on multiple screens
- for security clear passwords in memory after sent to backend
- (docu: pam_conv message flow in pam backend/helper/daemon/greeter)

qml notes:

- signals from daemon:
pamConvMsg(pam_msg, result) - provides infos/errors from (pam) backend conversation to present them to user in a info area, result from pam function for evaluation
pamRequest() - new request from (pam) backend, user response required

- new property:
request (type AuthRequest) - prompts from (pam) backend with messages from pam_conv

- response for backend:
sddm.enablePwdChange() - theme tells greeter it can handle expired passwords and change (dialog)
sddm.pamResponse(password) - send password response to (pam) backend, i.e. pam_conv
sddm.cancelPamConv() - cancel pam conversation

- new components:
PasswordChange.qml - dialog for conversation with pam for password change, may need some polishing
PamConvHelper.qml - hides pam  conversation handling for password change from themes, for Main.qml

sddm.conf settings:

LocaleFile - file path for distro-specific locale.conf file (for pam backend)
RetryLoop - ignore retry limit for pam_pwquality and loop pam_chauthtok on PAM_MAXTRIES

jean-yves-faye commented:
Works nicely here

jean-yves-faye commented on src/daemon/Display.cpp:
```diff
@@ -333,26 +356,41 @@ namespace SDDM {
+           info == Auth::INFO_PAM_CONV)
+        {
+            // send pam conversation message to greeter
+            emit pamConvMsg(m_socket, message);
```
Probably should check if m_socket is valid here, I've been having a crash with this stacktrace : 
`#0  0x00007f7b9afdcc3d in QIODevice::write(char const*, long long) ()
   from /usr/lib64/libQt5Core.so.5
#1  0x00000000004323ed in SDDM::SocketWriter::~SocketWriter() ()
#2  0x00000000004509fe in SDDM::SocketServer::pamConvMsg(QLocalSocket*, QString const&) ()
#3  0x00007f7b9b0b5831 in QMetaObject::activate(QObject*, int, int, void**) ()
   from /usr/lib64/libQt5Core.so.5
#4  0x0000000000455547 in SDDM::Display::pamConvMsg(QLocalSocket*, QString const&) ()
#5  0x000000000043d497 in SDDM::Display::slotAuthInfo(QString const&, SDDM::Auth::Info) ()
#6  0x00007f7b9b0b5831 in QMetaObject::activate(QObject*, int, int, void**) ()
   from /usr/lib64/libQt5Core.so.5
#7  0x0000000000455386 in SDDM::Auth::info(QString, SDDM::Auth::Info) ()
#8  0x0000000000434724 in SDDM::Auth::Private::dataPending() ()
`

It's caused by pam sending "creating directory" conv msg after auth success at first login

hoehnth replied:
hi, the changes for that were already laying around for some time, just had to transfer them to git. is done in the next 3 commits. just have to test them in my new build.

hoehnth commented:
hi david,

would have been nice if the patches were reviewed earlier. now i have to merge again for new 0.15.0 version...

with the detailed notes earlier i wanna make sure that my implementation with password propagation through greeter/daemon/backend is basicly ok for you maintainers (other option was password renewal directly in greeter but i am out to re-implement this).

- themes:
you talk about the changes to maui (default)/elarun/maldives themes?
well my intention was just to provide a convenient helper with the PasswordConnections.qml which hides/connects common properties/methods for password renewal. i prepared the sddm stock themes to be ready to use and to serve as example how to do password renewal. otherwise the same redundant code from  PasswordConnections.qml goes into all themes, ugly no?! its just an offer to minimize work for others to use password renewal in their themes.

i dont understand what you mean with themeAPI in desktop file...

- SDDM_BACKEND define:
as written in AuthRequest.h comment i would prefer to disable any qml related stuff in the backend (helper, backend, pam backend) to keep objects there clean, qml props have no use there and are just ballast. the qml props make only sense in the greeter frontend. bus its just a minor thing, can be removed if not ok.

regards,
thomas

davidedmundson commented:
>would have been nice if the patches were reviewed earlier. now i have to merge again for new 0.15.0 version...

Yeah, I know...I had one on here for over a year waiting for a review. I fully understand :/ 

But I'm taking over to try and fix all that now.

>with the detailed notes earlier i wanna make sure that my implementation with password propagation through greeter/daemon/backend is basicly ok for you maintainers (other option was password renewal directly in greeter but i am out to re-implement this).

Yeah, I think it makes sense. It's what LightDM does too.

>themes:

What you've done in the QML is fine.

My concern is, pretend I'm running an old theme and then we ship SDDM with this
We get a PAM prompt, we send a message to the greeter, the greeter ignores it, the user session just sits there waiting for me to reply.

Can that happen? and if so, is it different to the current state when we get a PAM prompt?

> qml props have no use there

They're not "QML props" they're standard Qt metaobject properties; we had them before QML. I don't think there's any need to complicate it. And if we do ever want more stuff in the QML exposed object - we have subclasses or qmlRegisterExtendedType

hoehnth commented:
rebased branch to current sddm version 0.15.0...
just an update to src/greeter/CMakeLists.txt was needed so far.

davidedmundson commented:
You haven't addressed the only thing I asked about.

Can you give me some instructions on how you're testing expiring a password. Then I can check and if need be fix it.

jean-yves-faye commented:
To expire an user, use chage -d 0 <user>

hoehnth commented:
hi david, please give me some time... i was going to answer your questions right next after testing here.
i just wanted finish updating to new sddm version first.

well, you are right, with an old theme the (pam) backend just waits for a reply (from greeter) with my version. so not yet ready for merging (we dont have that problem with older themes so I didnt prepare yet). 

in current sddm version the password expired pam messages are just sent to greeter as login failure.

in order to support older themes:
greeter can ask theme for a property (only set in new themes) and cancel pam conversation if not found. that would be easy to implement.

btw to expire user password, you can also use:
passwd -e user

hoehnth commented:
hi, now you can test with old and new themes.

hoehnth commented:
do you see any other point to improve?

should split the changes for AuthBase so rest is easier to digest?

davidedmundson commented:
Thanks, wasn't trying to rush you earlier.

> for themes which do not have password renewal dialog the pam conversation is just canceled.

That's absolutely fine. 

>should split the changes for AuthBase so rest is easier to digest?
I think it's fine as-is. 

I'll have some time Monday to properly review and test it, but I don't forsee any major changes.

hoehnth commented:
> Thanks, wasn't trying to rush you earlier.
ok, no problem.

> I'll have some time Monday to properly review and test it, but I don't forsee any major changes.
would be good. i don't have any more modifications to this PR so far. just some minor changes.
have added short doc which might help other themes.
and last, i could prepare maya theme.

davidedmundson commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+    property var pwdItem: QtObject
+
+    // if provided supposed to have text property
+    property var errMsg: QtObject
```
I don't understand why we're creating a default QtObject if unset.

hoehnth replied:
ok, then initialize as undefined is maybe better.

davidedmundson commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+/***************************************************************************
```
I'm not super convinced by some of this class. 
We're creating a template style approach (which could be fine), but only for one minor feature. 

it's just as easy to add:

enabled: !renewalDialog.visible into the theme in the right places.
as it is to pass a list of items in here.

hoehnth replied:
sure, we can also set enabled in the themes to block the item(s). the disabledItems was though just as a helper to collect/bundle the logic for password renewal. i can kick it out if you dont like it.

hoehnth commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+
+Item {
+
+    property var sddmProxy: QtObject
```
thanks for pointing that out. wasn't aware of the root context meaning before. is updated.

davidedmundson replied:
This page is worth reading http://doc.qt.io/qt-5/qtqml-documents-scope.html

hoehnth commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+        onLoginFailed: {
+            if(txtMsg.objName !== "")
+            {
+                txtMsg.color = "red"
```
i agree. will change that.

hoehnth commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+
+        // new pam request arrived, e.g. for expired password,
+        onPamRequest: {
+            // open password renewalDialog dialog and block other GUI
```
if it is such a resource concern yes, will change it later.

hoehnth commented on components/2.0/PasswordConnections.qml:
```diff
@@ -0,0 +1,151 @@
+                if(err_msg != "Authentication failure" &&
+                   err_msg != "Password change aborted." &&
+                   err_msg != "Authentication token manipulation error")
+                   txtMsg.text += "\n" + err_msg
```
first:
> There's a very important comment about reading PAM data and i18n. That needs fixing at a core level.

well, i agree the code should be ready for i18n setups.
if we have to handle multi-language messages from PAM itself, then the PamBackend.cpp has first to be amended. it only handles ascii messages currently. i based my changes on that version to extend it and use those messages in the frontend. unfortunately pam conversation gives no kind of message ID to easily identify messages but just msg_style and raw string which is not enough. at the moment the type is just found my string comparison. who will do that i18n extensions the to pam backend? as we need this here to maybe i can implement, but not sure if i have time soon.

> I've had this before when doing multiple auth prompts. Probably best done in the C++ code rather than the QML.

we can do some preprocessing outside of qml, sure.

plfiorini commented:
@hoehnth and @davidedmundson what is the status of this PR?

hoehnth commented:
I did not have time yet to get the (PAM) backend ready for i18n.
Was busy with other projects at work so far.

But now seems possible for me to start that work in the next 1 or 2 weeks.
Will let you know when work is in progress.

hoehnth commented:
Working on it. May take some time...
I had a look in gdm3 sources how its done there.
Next, separated helper out of sddm to have a standalone test program for authentication.
Now trying to get rid of message string comparisons.
I better don't promise anything yet...

hoehnth commented:
Lots of changes in the (pam) backend...
Integrated changes from backend test program in sddm backend.
Updated greeter and daemon for backend changes.
Now testing frontend & updating expired password dialog.
Next, have to transfer changes from our sources into pull-request.

hoehnth commented:
@davidedmundson Updated the themes Main.qml and removed some settings from the former PasswordConnections.qml template style file you didn't like, now renamed PamConvHelper.qml - handles only PAM conv related logic now.

piotr-dobrogost commented on src/greeter/theme/Main.qml:
```diff
@@ -110,6 +110,8 @@ Rectangle {
         Row {
             anchors.fill: parent
+            // hide main input for small displays when password change dialog is active
+            visible: !(passwordChange.visible && container.height < 768)
```
Shouldn't break point be a param instead of hard coded here and elsewhere?

hoehnth replied:
Sure you can put that size value in a param, but I think its not so important at the moment.
If the pull request ever gets accepted - and thats the question - you are free to change it as you like. Someone might also completely redesign the pwd change dialog to fit in the theme, or overlay it, I don't mind.

davidedmundson commented:
Thanks for tackling that localisation issue.

I'm going to do a release today, then I'll spend time looking at the new features and having a merging spree

hoehnth commented:
A few minor improvements still left (pwd change dialog in maldives and elarun theme).

hoehnth commented:
Hi, I think there is no need to merge develop into my pull-request. Will rebase PR on develop branch from time to time.

Overc1ocker commented:
What is the status of this PR?

hoehnth commented:
> What is the status of this PR?

Probably same status & fate like other PR's - not going to be merged because no one has the time or interest to review.
But I also know this patch is huge with lots of changes to the backend so maybe no one dares to merge ;-)

hhl commented:
Hi, i am working on a translation to german (de) and I stumbled over this phrase  see 
https://github.com/sddm/sddm/pull/1121#issuecomment-453502616 and https://github.com/sddm/sddm/pull/1121#issuecomment-453828811.
What do you want to say with it?
See https://github.com/sddm/sddm/pull/1121#discussion_r329928043
I think, in the end I would change


```qml
readonly property string emptyPrompt:       qsTr("Empty prompt from pam")
```
into
```qml
readonly property string emptyPrompt:       qsTr("unregistered user")
```
 I reference to
components/common/TextConstants.qml -> line 40 
( https://github.com/sddm/sddm/pull/776/files#diff-2aea672f24105b1f754f3a6b49f9d977 )
wich is linkt to 
components/2.0/PasswordChange.qml -> line 94 - 102 
( https://github.com/sddm/sddm/pull/776/files#diff-61ab0c3336a43d7432e2d789abf89c20 )
(sorry don't know how to copy the diff parts here) 

So, what shall write?

hoehnth commented:
> @hoehnth
> Hi, i am working on a translation to german (de) and I stumbled over this phrase
> 
> ```qml
> readonly property string emptyPrompt:       qsTr("Empty prompt from pam")
> ```
> (sorry don't know how to copy the diff parts here)
> So, what shall I write?

The emptyPrompt is for the unlikely (error) case, that the PAM module (pam_chauthtok, called by pam_authentication, see man page) sends an empty prompt during authentication conversation. I don’t think it will ever happen, because all prompts are well defined. The prompt, like “Password:” or “Repeat password:” will be shown as label in front of the input field for the user to enter a (new) password. So in an error case of an empty prompt I should rather exit the password change dialog with the error string “Empty prompt from pam”. A text like “unregistered user” should not be used. Better use “Password:”, because prompts in the PAM conversation always expect some kind of password input.

SolarAquarion commented:
can you rebase the patch?

hoehnth commented:
i can do when i have time again.
but to be honest, its only worth for me to invest work again on this patch, when its considered to be merged. it could have be done already 2 years ago. its not my fault when the patch gets out of sync because its not merged and devel on main trunk continues. so better address or push the maintainers if you consider the patch important.

pguermo commented:
Hello @SolarAquarion, @davidedmundson, 

SDDM is currently not usable in an enterprise environment where password may be forced to be changed. So this fix is mandatory for a such use case... otherwise we have to use gdm or kdm as fallback solution.

I understand the frustration of @hoehnth if he did several rebases as the review and merge was not done. Could you commit yourself to review the issue and merge it in the following days then once the issue is one more time rebased ? Thanks in advance.

hoehnth commented:
> SDDM is currently not usable in an enterprise environment where password may be forced to be changed. So this fix is mandatory for a such use case... otherwise we have to use gdm or kdm as fallback solution.
> 
hi pguermo, if you need this feature urgently for your enterprise environment, i may able to provide a rebased patch for the latest sddm version on demand. you can contact me in case you consider it seriously. otherwise i will just wait with rebase until the PR is really considered for merging.

hoehnth commented:
> Hello @SolarAquarion, @davidedmundson,
> 
> SDDM is currently not usable in an enterprise environment where password may be forced to be changed. So this fix is mandatory for a such use case... otherwise we have to use gdm or kdm as fallback solution.
> 
> I understand the frustration of @hoehnth if he did several rebases as the review and merge was not done. Could you commit yourself to review the issue and merge it in the following days then once the issue is one more time rebased ? Thanks in advance.

pguermo commented:
Hi @hoehnth. Thanks for your reaction. As this feature is currently not in present in SDDM, we CANNOT go in production using sddm. I cannot understand why the project manager of sddm doesn't give a higer priority of this issue (but it's not too late guys ;) ). 

Thanks for your proposal Thomas. I really appreciate it ! I hope good willing reviewers will then act and the patch will be embeded officially in our favorite distros.

hoehnth commented:
Maybe closed it accidentally...

chasonr commented:
I can rebase this change. I have this change, along with other small changes by @etaylorquoin, working with the 0.19.0 release.

RobVerduijn commented:
I've been waiting for this to be merged since 2017, I would definitely appreciate it.

pguermo commented:
Great ! Thanks in advance @chasonr

roberthoffmann commented on src/daemon/Display.cpp:
```diff
@@ -370,52 +396,120 @@ namespace SDDM {
+             return;
         }
+
+        qWarning() << "Display: Unable to handle Auth request!";
```
Cancel auth requests which cannot be handled. Otherwise clicking the Login button with an empty username field would prevent subsequent login attempts:
```
m_auth->request()->canceled();
````

hoehnth replied:
Sure, makes sense. I can add it.

hoehnth replied:
Done.

hoehnth commented:
Rebased to v0.19.0

Update:
Rebase saved to branch expired-password-v0.19.0

Iniesta8 commented:
@davidedmundson @plfiorini : What is the status of this PR? Is there any work to do before it can be merged?

It would be great if this functionality comes into sddm.

plfiorini commented:
Ping.

Is this PR ready to be reviewed?
If yes, can you please resolve the conflicts and squash all those commits?

hoehnth commented:
@Iniesta8 @plfiorini All commits squashed and rebased to develop branch.

hoehnth commented:
@plfiorini @davidedmundson PR ready for review.

pguermo commented:
Thanks @hoehnth ! 

What is this issue waiting for beeing merged ?

hoehnth commented:
> Thanks @hoehnth !
> 
> What is this issue waiting for beeing merged ?

@davidedmundson , @plfiorini Hi, anything which needs to be improved on the PR? Can you answer @pguermo?

pguermo commented:
No !!! One more time conflicts occured as no one reviewed on time this issue... 

What is making @aleixpol, @Vogtinator or @davidedmundson so afraid to review this change ?

@plfiorini What's the next step for this issue? What's the future of SDDM? Without this feature, SDDM may not be used in a corporate network where users must be notified their password will expire soon or has expired. This is definitely not a cosmetic bug.

0n0w1c commented:
We continue to use lightdm and slick-greeter to resolve this business need.

julietrb1 commented:
I have to add my two cents here: I'm rolling out an initiative to bring 40-50 computers to "polished" small-business level by adding FreeIPA, and I am precluded from using KDE because of this.

Vogtinator commented:
> I have to add my two cents here: I'm rolling out an initiative to bring 40-50 computers to "polished" small-business level by adding FreeIPA, and I am precluded from using KDE because of this.

Unlike some other desktop environments, Plasma does not require the use of a specific display manager.

hhl commented:
I really want to know where the problem is to integrate that little piece of code which add a useful feature to sddm.
2017 -> 2021 wow!!!!

RiderALT commented:
We (altlinux) have long been, about 3 years we use sddm with this patch for corporate use in the production for thousands of workstations.

Mazino-Urek commented:
@davidedmundson @aleixpol @plfiorini Kindly have a look at this.

Pointedstick commented:
This needs the merge conflicts fixed before anyone can review it. @hoehnth are you up for that? If you can do that, I'll try to get some attention on this pull request so it can eventually be merged.

Mc-Pain commented:
> This needs the merge conflicts fixed before anyone can review it. @hoehnth are you up for that? If you can do that, I'll try to get some attention on this pull request so it can eventually be merged.

I did it here: https://github.com/Mc-Pain/sddm/tree/expired-password

hoehnth commented:
I'll do a rebase when I have time. But I doubt it will kick off review / merge this time.

@Mc-Pain I would have to review your rebase to make sure nothing is missing, but I think will take same time doing rebase myself so I prefer own rebase.

Mc-Pain commented:
> I'll do a rebase when I have time. But I doubt it will kick off review / merge this time.
> 
> @Mc-Pain I would have to review your rebase to make sure nothing is missing, but I think will take same time doing rebase myself so I prefer own rebase.

I think reviewing conflicted files would be nice and quicker than reviewing the whole rebase.

**UPD** or, at least, build with rebased patch and test for any obvious bugs.

Mazino-Urek commented:
@hoehnth Nate is in KDE development team. This time it has a better chance to be reviewed and merged.

hoehnth commented:
rebased. short test looks ok. have to check commit again though if all looks ok.

Mc-Pain commented:
It's diverged again. Pushed 00:55 GMT+3, merged #1449 01:33 GMT+2

hoehnth commented:
> It's diverged again. Pushed 00:55 GMT+3, merged #1449 01:33 GMT+2

no panic. i'll fix that soon.

Mazino-Urek commented:
@Pointedstick ping.

0n0w1c commented:
Maybe the pull request is lacking, needing to address this?
```
- SDDM is a modern display manager for X11 and Wayland aiming to be fast, simple and beautiful.
+ SDDM is a modern display manager for X11 and Wayland aiming to be fast, simple, beautiful and fully functional.
```

Pointedstick commented:
Thanks, I will try to get this reviewed by someone.

Mc-Pain commented:
Aaaaaand...
No review, branches are diverged again, 3+ weeks passed.

@Pointedstick

Pointedstick commented:
People are just busy, and patience is requested.

davidedmundson commented:
>People are just busy, and patience is requested.
 
We can't say that after you explicitly asked him to rebase it, that's just aggressively wasting other people's time.

Don't rebase again, I'll sort it.

Mc-Pain commented:
> > People are just busy, and patience is requested.
> 
> We can't say that after you explicitly asked him to rebase it, that's just aggressively wasting other people's time.
> 
> Don't rebase again, I'll sort it.

plz tell us when it will be sorted and reviewed

0n0w1c commented:
This PR has been here for over 4 years, don't hold your breath.

ryzendew commented:
so ... is anyone left alive to finish and merge this PR

now-im commented:
> so ... is anyone left alive to finish and merge this PR

Unless someone forking this mess, "NOPE".

RobVerduijn commented:
it seems I completely forgot the 5th anniversary of this pull request
I'm sorry, I really wanted to do something special, can I make it up with some flowers ?

Mc-Pain commented:
ping

rferris commented:
![image](https://user-images.githubusercontent.com/1245687/198208894-84602447-0c1d-4b23-a211-54a2bb4ce59b.png)

Mc-Pain commented:
![изображение](https://user-images.githubusercontent.com/9453287/198216718-1c90fe2e-aa07-4ecf-886b-6d441edae1f3.png)

wschlich commented:
sigh...

aleixpol commented:
I understand it's kind of fun to make memes about PRs and stuff, but it's also worth noting that it's a hard one to take in.

* It has had conflicts for at least 18 months.
* It adds 2000 lines of code in fairly untested sensitive code areas.
* The original author is understandably worried about other parts in the meantime.
* It solves a use-case that is alien to many of the current SDDM devs.
* It changes front-end code and will potentially break currently existing themes.

How to get this merged? To be honest I am not sure. Making sure that people has tested this and ensures this addresses all their problems would be a great first step. Doing so with the different themes or at least laying out how themes (e.g. Breeze's) ought to be changed, would also be very useful.

Mc-Pain commented:
> * It has had conflicts for at least 18 months.

Which had appeared at least three times after this PR has not been reviewed in time, causing other PRs being merged.

> * It adds 2000 lines of code in fairly untested sensitive code areas.

The more code involved - more likely other PRs would cause a merge conflict.

First, someone should rebase it onto current and review/test ASAP.
Either @davidedmundson, as pointed out [here](https://github.com/sddm/sddm/pull/776#issuecomment-976100829)
Or @hoehnth, prefering [to rebase instead of review](https://github.com/sddm/sddm/pull/776#issuecomment-945795086)

aleixpol commented:
Here's this PR rebased on origin/develop: https://github.com/sddm/sddm/tree/expired-password

I can create a PR from it if that makes anyone's life any easier.
If somebody wants to help, splitting  this into a few separate commits would be handy. It's called "expired passwords" and "pam message localization" which are two entirely different topics.

hoehnth commented:
Sorry to say, but honestly i lost most interest to spend more time on this PR regarding the past history with the rebase and wait game. It sounds like a joke to me it had conflicts for xy months...
Sure, the best to do for merge would be a split in separate commits. I offered that a long time ago when i had time, and was willing to do it. I'm full time programmer, like other developers, and now have other things to do after work... Maybe when I'm bored on a weekend I may consider it.
I provided rebased branches for (0.14.0,) 0.18.0, 0.18.1, 0.19.0 releases which gave enough time for testing. I dont see the benefit of another rebase at the moment. I think for testing before merging to develop it should be kept in a testing branch and add the commits there.
The use-case of the PR may be alien to many sddm devs and home users, but i18n/localization for pam and password renewal is a must if used with sssd setup and active-directory in (international) company environment, as companies usually have password renewal schemes with AD or LDAP. For this purpose the original code was written. It adds code in sensitive code areas? Well, if someone can provide a simpler solution, I don't mind, skip this PR, I dont need it...
Luckily - I thought - my company allowed me to share the (productive) code with oss community, so we can contribute and give back, but it seemed it is not valued as much as I thought. Though this is just a small mod for sddm we have done to customize it...
Btw, its already documented how other themes ought to use the feature in EXPIRED-PASSWORD.md
It's fully documented already.

PS: Nice memes!

hoehnth commented:
Had some time and rebased last week, but it's untested yet.
No need to review as I will split up the PR some day.

rnayabed commented:
I'll probably die before this PR ever gets merged

Overc1ocker commented:
> I'll probably die before this PR ever gets merged

This is unhelpful and makes the PR look unprofessional. Please be professional and helpful in your comments.

rnayabed commented:
My comment hardly makes a difference. What is unprofessional is maintainers repeatedly saying they'll look and merge this then go away. It's perfectly ok since its community work, so understandable. Then, why don't they just step down?

xBr0th3rx commented:
Is there a timeline for when the feature will be looked at and implemented?

Mc-Pain commented:
ping

chart41 commented:
Hello from 2k24.
I installed it in Kubuntu. And it looks not bad, but not quite finished yet.

1. Works only on default theme, which not showed in settings (there is just no selected theme)
2. After start you can see on-screen keyboard, which covers everithing, and you can't see nothing before you disable it (down-right button)
3. If password wrong, you won't see any message, the field just become empty again.
4. If password expiried, you will see new modal windows where you can enter current pass and then new (windows for each input).

**So, I think we need some new hero programmer, who can take care about this options:**
(Wonder how many years of development of this standard regular feature might take)

1. (which most important for me, and obviously most simplest) Add wrong password message (and locale).
2. You can use this features in any theme
3. Add on-screen keyboard button, so the keyboard appers when this button clicked, and doesn't cover everithing
4. You can see this theme in themes list in settings.
5. It is not necessary to create new window for every new password field, you can show this fields one by one on the main layer, and maybe show additional buttons for them.

_If we get this option, maybe more companies will be interested in using Kubuntu in domain systems._

**How I installed it in Kubuntu 22.04 or 23:**
```
sudo apt purge sddm
sudo apt upgrade
sudo apt-get install cmake extra-cmake-modules libpam0g-dev libsystemd-dev libupower-glib-dev libxcb-xkb-dev libxcb1-dev libxkbcommon-dev pkg-config python3-docutils qtbase5-dev qtdeclarative5-dev qttools5-dev qttools5-dev-tools build-essential
git clone https://github.com/hoehnth/sddm
cd sddm
git checkout expired-password
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX=/usr
make
sudo make install
sudo apt-get remove cmake extra-cmake-modules libpam0g-dev libsystemd-dev libupower-glib-dev libxcb-xkb-dev libxcb1-dev libxkbcommon-dev pkg-config python3-docutils qtbase5-dev qtdeclarative5-dev qttools5-dev qttools5-dev-tools build-essential
sudo systemctl enable --now sddm
```

jbruechert commented:
Since there was [some interest on the discuss.kde.org forum](https://discuss.kde.org/t/sddm-doesnt-support-expired-password-prompt-ad-account/26006), I wondered how much work this was to rebase.
I ended up with something that at least compiled pretty easily.
So in case someone wants to pick this up again, you can start from my rebased branch: https://github.com/jbruechert/sddm/tree/expired-password

petterreinholdtsen commented:
[Jonah Brüchert]
> I ended up with something that at least compiled pretty easily.
> So in case someone wants to pick this up again, you can start from my
> rebased branch:
> https://github.com/jbruechert/sddm/tree/expired-password

It make me glad to see someone working on this again.  Would be great if
sddm could be brought up to par with Windows and gdm in this regard. :)

-- 
Happy hacking
Petter Reinholdtsen

Naming-things-is-hard-btw commented:
sheesh what a sad story
reminds me of my ancient forgotten PR's i made for godot (with a different account) that never got merged

YamiYukiSenpai commented:
> Since there was [some interest on the discuss.kde.org forum](https://discuss.kde.org/t/sddm-doesnt-support-expired-password-prompt-ad-account/26006), I wondered how much work this was to rebase. I ended up with something that at least compiled pretty easily. So in case someone wants to pick this up again, you can start from my rebased branch: https://github.com/jbruechert/sddm/tree/expired-password

Would it be possible for this to compile on Kubuntu 24.04?  I've never packaged stuff before and looking into doing this for work and add it on our local repo.

jbruechert commented:
> Would it be possible for this to compile on Kubuntu 24.04? I've never packaged stuff before and looking into doing this for work and add it on our local repo.

It should be possible depending on which version of the dependencies Kubuntu 24.04 is shipping.

However I'm not sure I would deploy the changes as they are. Rebases can always introduce new issues.

I'm pretty sure the reason why this doesn't progress is that this change touches the core of SDDM and is very big (dangerous + hard to review) . Additionally, it is usually not needed for private setups (relatively low interest from volunteer develoeprs).
So if a business needs this and has the resources, consider spending some resources to have someone split out the individual changes that the final feature depends on, like "localization (i18n) for pam messsages and prompts".

SgAkErRu commented:
Perhaps it will be interesting to someone. 

I have created a heavily modified and simplified implementation of this feature based on this PR.

The idea was to implement this with minimal changes to the existing sddm architecture and code.

The implementation has been split into the commits and rebased to the latest SDDM state (as of today) of the develop branch: 

https://github.com/SgAkErRu/sddm/commits/develop-expired-password/

A small patch (without new components, based on reuse of existing ones) has been implemented separately to support this functionality for Plasma 6.4 Breeze Sddm theme here:

https://invent.kde.org/sgakerru/plasma-desktop/-/commits/expired-password-6.4
